ncbi/src/main/java/org/reactome/release/PathwayHierarchyUtilities.java

```diff
@@ -0,0 +1,203 @@
+package org.reactome.release;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.neo4j.driver.v1.Record;
+import org.neo4j.driver.v1.Session;
+import org.neo4j.driver.v1.StatementResult;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Utilities class for retrieving Reactome Pathway Hierarchy information. This class has logic
+ * for querying relationships between Reaction Like Events and their Pathways, child and parent
+ * Pathways, and Top Level Pathways in Reactome.
+ * @author jweiser
+ */
+public class PathwayHierarchyUtilities {
+	private static final Logger logger = LogManager.getLogger();
+
+	private static Map<Session, Map<Long, Set<Long>>> rleToPathwayIdCache = new HashMap<>();
+	private static Map<Session, Map<Long, Set<Long>>> pathwayHierarchyCache = new HashMap<>();
+	private static Map<Session, Set<Long>> topLevelPathwayIdsCache = new HashMap<>();
+
+	/**
+	 * Retrieves, from the graph database, a Map of Reaction Like Event database identifiers
+	 * to the Set of Pathway database identifiers in which each Reaction Like Event participates
+	 * @param graphDBSession Neo4J Driver Session object for querying the graph database
+	 * @return Map of Reaction Like Event identifier to Set of Pathway identifiers
+	 */
+	public static Map<Long, Set<Long>> fetchRLEIdToPathwayId(Session graphDBSession) {
+		if (rleToPathwayIdCache.containsKey(graphDBSession)) {
+			return rleToPathwayIdCache.get(graphDBSession);
+		}
+
+		if (graphDBSession == null) {
```

Comment:
Can null be used as a key to a Map? Would it make more sense to have this null check before the above `containsKey` condition?

Reply:
I am pretty sure that a Java Map can have a key that is null.

Reply:
You're right -- the null check should be done first.  I've changed the order.